nodeadm/internal/containerd/config.go

```diff
@@ -66,8 +100,41 @@ func generateContainerdConfig(cfg *api.NodeConfig) ([]byte, error) {
 		EnableCDI:         semver.Compare(cfg.Status.KubeletVersion, "v1.32.0") >= 0,
 	}
 	var buf bytes.Buffer
+	containerdConfigTemplate := template.Must(template.New(containerdConfigFile).Parse(containerdTemplateVersionMap[templateVersion]))
 	if err := containerdConfigTemplate.Execute(&buf, configVars); err != nil {
 		return nil, err
 	}
 	return buf.Bytes(), nil
 }
+
+func getConfigTemplateVersion(cfg *api.NodeConfig, isContainerdV2 bool) (TemplateVersion, error) {
+	if isContainerdV2 {
+		// side case: if v2 config passed in nodeConfig when using containerd 2.*, we use V2 config template and will run containerd config migrate
+		if len(cfg.Spec.Containerd.Config) > 0 && !Version3configInNodeConfig(cfg) {
+			return V2, nil
+		}
+		return V3, nil
+	} else {
+		// side case: if v3 config passed in nodeConfig when using containerd 1.*, throw error
+		if len(cfg.Spec.Containerd.Config) > 0 && Version3configInNodeConfig(cfg) {
+			zap.L().Error("Invalid containerd config passed, containerd 1.7.* doesn't support containerd configuration V3 properties")
+			return "", fmt.Errorf("failed to get config template version")
+		}
+		return V2, nil
+	}
+}
+
+// Most proprty are moved under `io.containerd.cri.v1.runtime` and `io.containerd.cri.v1.images` in config v3, only a few left
+// in `io.containerd.grpc.v1.cri` which is same as config V2. So assume it is config V2 if `io.containerd.grpc.v1.cri` detected
+func Version3configInNodeConfig(cfg *api.NodeConfig) bool {
```

Comment:
nit: maybe change the signature to something like
```suggestion
func isConfigSchemaV3(config string) bool {
```

Reply:
revised